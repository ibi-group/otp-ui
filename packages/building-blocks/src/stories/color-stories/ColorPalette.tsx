import React, { useEffect, useState } from "react";

import styled, { css } from "styled-components";
import { getMostReadableTextColor } from "@opentripplanner/core-utils/lib/route";
import WcagChecker from "./WCAGChecker.story";
import ColorPickerInput from "./ColorPickerInput.story";
import blue from "../../colors/blue";
import grey from "../../colors/grey";
import SelectedColorOptions from "./SelectedColorOptions.story";
import { Hue } from "../../types/hue";

interface Props {
  color: Hue;
  label: string;
}

export const borderStyles = css`
  border: 1px solid ${grey[100]};
  border-radius: 7px;
`;

const Info = styled.div`
  background-color: ${blue[50]};
  line-height: 1.5;
  padding: 10px;
  border-radius: 5px;
  margin-bottom: 15px;
  text-align: center;

  p {
    margin: 0;
  }

  a {
    color: ${blue[700]};
    font-weight: 700;
    text-decoration: none;

    &:hover {
      color: ${blue[900]};
    }
  }
`;

const ColorContainer = styled.div`
  display: flex;
  gap: 15px;
  font-family: Arial, Helvetica, sans-serif;
  margin: 0 auto;
`;

const ColorPaletteContainer = styled.div``;

const ColorBlock = styled.div<{ hex: string }>`
  align-items: center;
  background-color: ${props => props.hex};
  color: ${props => getMostReadableTextColor(props.hex, "#ffffff")};
  text-shadow: ${props =>
    getMostReadableTextColor(props.hex, "#ffffff") === "#ffffff"
      ? "1px 1px 2px black"
      : ""};
  display: flex;
  height: 40px;
  justify-content: space-between;
  margin: 0;
  padding: 10px;
  transition: all 0.2s ease-in;
  width: 300px;

  &:first-of-type {
    border-radius: 8px 8px 0 0;
  }
  &:last-of-type {
    border-radius: 0 0 8px 8px;
  }

  &:hover {
    transform: scale(1.012);
    transition: all 0.2s ease-in;
    cursor: pointer;
  }
`;

const ColorPickForm = styled.form`
  display: flex;
  gap: 15px;
  margin-bottom: 15px;
`;

const WCAGContainer = styled.div`
  max-width: 425px;
`;

const ColorPalette = ({ color, label }: Props): JSX.Element => {
  const colorArray: Array<Hue> = Object.entries(color);
  const [selectedColor, setSelectedColor] = useState(colorArray[0]);
  const [foregroundColor, setForegroundColor] = useState(colorArray[0][1]);
  const [foregroundColorInput, setForegroundColorInput] = useState(
    colorArray[0][1]
  );
  const [backgroundColorInput, setBackgroundColorInput] = useState("#FFFFFF");
  const [backgroundColor, setBackgroundColor] = useState("#FFFFFF");
  const [foregroundError, setForegroundError] = useState(false);
  const [backgroundError, setBackgroundError] = useState(false);

  const REX_EXP = /(^#[0-9A-F]{6}$)/i;

  const isValidHexCode = code => {
    let hexcode = code;
    if (!code.startsWith("#")) hexcode = `#${code}`;
    return REX_EXP.test(hexcode);
  };

  const backgroundHandleChange = event => {
    const value = event.target.value;
    setBackgroundColorInput(value);
  };

  const foregroundHandleChange = event => {
    const value = event.target.value;
    setForegroundColorInput(value);
  };
  const hexArray = colorArray.map(x => x[1]);
  const foregroundColorIsInPalette = () => {
    if (foregroundColorInput && hexArray.includes(foregroundColorInput)) {
      return `${label} ${
        colorArray[hexArray.indexOf(foregroundColorInput)][0]
      }`;
    }
    return null;
  };
  const backgroundColorIsInPalette = () => {
    if (backgroundColorInput && hexArray.includes(backgroundColorInput)) {
      return `${label} ${
        colorArray[hexArray.indexOf(backgroundColorInput)][0]
      }`;
    }
    return null;
  };
  const selectedColorReset =
    selectedColor &&
    foregroundColorInput !== selectedColor[1] &&
    backgroundColorInput !== selectedColor[1];

  useEffect(() => {
    if (selectedColorReset) {
      setSelectedColor(undefined);
    }
    if (!isValidHexCode(foregroundColorInput)) {
      setForegroundError(true);
    } else {
      setForegroundError(false);
      setForegroundColor(foregroundColorInput);
    }
  }, [foregroundColorInput]);

  useEffect(() => {
    if (selectedColorReset) {
      setSelectedColor(undefined);
    }
    if (!isValidHexCode(backgroundColorInput)) {
      setBackgroundError(true);
    } else {
      setBackgroundError(false);
      setBackgroundColor(backgroundColorInput);
    }
  }, [backgroundColorInput]);

  return (
    <ColorContainer>
      <ColorPaletteContainer>
        {colorArray.map(
          (hue: Hue): JSX.Element => {
            return (
              <ColorBlock
                key={hue[0]}
                hex={hue[1]}
                onClick={() => setSelectedColor(hue)}
              >
                <p>{hue[0]}</p>
                <p>{hue[1]}</p>
              </ColorBlock>
            );
          }
        )}
      </ColorPaletteContainer>

      <WCAGContainer>
        <Info>
          <p>
            Select a color variable from the palette to check it against WCAG
            contrast requirements.{" "}
            <a
              target="_blank"
              rel="noopener noreferrer"
              href="https://webaim.org/articles/contrast/"
            >
              See more
            </a>{" "}
          </p>
        </Info>
        <SelectedColorOptions
          selectedColor={selectedColor}
          selectedColorLabel={label}
          setBackgroundColor={setBackgroundColor}
          setForegroundColor={setForegroundColor}
          setBackgroundColorInput={setBackgroundColorInput}
          setForegroundColorInput={setForegroundColorInput}
        />
        <ColorPickForm>
          <ColorPickerInput
            colorInPalette={foregroundColorIsInPalette()}
            handleChange={foregroundHandleChange}
            inputLabel="Foreground color"
            colorInput={foregroundColorInput}
            error={foregroundError}
          />
          <ColorPickerInput
            colorInPalette={backgroundColorIsInPalette()}
            handleChange={backgroundHandleChange}
            inputLabel="Background color"
            colorInput={backgroundColorInput}
            error={backgroundError}
          />
        </ColorPickForm>
        <WcagChecker background={backgroundColor} hue={foregroundColor} />
      </WCAGContainer>
    </ColorContainer>
  );
};
export default ColorPalette;
