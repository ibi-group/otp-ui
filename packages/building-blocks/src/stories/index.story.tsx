import React, { ReactElement } from "react";
import { Meta } from "@storybook/react";
import ColorPalette from "./ColorPalette";

import grey from "../colors/grey";
import blue from "../colors/blue";
import red from "../colors/red";

const meta: Meta<typeof ColorPalette> = {
  title: "Building-Blocks/Colors",
  component: ColorPalette
};

export default meta;

export const Grey = (): ReactElement => {
  return <ColorPalette color={grey} />;
};

export const Blue = (): ReactElement => {
  return <ColorPalette color={blue} />;
};

export const Red = (): ReactElement => {
  return <ColorPalette color={red} />;
};
