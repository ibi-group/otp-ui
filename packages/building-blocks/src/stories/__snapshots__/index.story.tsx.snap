// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Building-Blocks/Colors Blue smoke-test 1`] = `
<div class="sc-aXZVg laHUje">
  <p>
    50
  </p>
  <p>
    #E5ECF3
  </p>
</div>
<div class="sc-aXZVg hCUOdW">
  <p>
    100
  </p>
  <p>
    #CCDAE7
  </p>
</div>
<div class="sc-aXZVg fvCtxD">
  <p>
    200
  </p>
  <p>
    #B2C7DB
  </p>
</div>
<div class="sc-aXZVg hYYmfL">
  <p>
    300
  </p>
  <p>
    #99B5CF
  </p>
</div>
<div class="sc-aXZVg iRVeSf">
  <p>
    400
  </p>
  <p>
    #7FA2C2
  </p>
</div>
<div class="sc-aXZVg liHIEc">
  <p>
    500
  </p>
  <p>
    #6690B7
  </p>
</div>
<div class="sc-aXZVg euPFdB">
  <p>
    600
  </p>
  <p>
    #4D7EAB
  </p>
</div>
<div class="sc-aXZVg hMMvZc">
  <p>
    700
  </p>
  <p>
    #336B9E
  </p>
</div>
<div class="sc-aXZVg fEXWvT">
  <p>
    800
  </p>
  <p>
    #1A5992
  </p>
</div>
<div class="sc-aXZVg kjzUak">
  <p>
    900
  </p>
  <p>
    #004686
  </p>
</div>
`;

exports[`Building-Blocks/Colors Grey smoke-test 1`] = `
<div class="sc-aXZVg bZDWEL">
  <p>
    50
  </p>
  <p>
    #ECECEC
  </p>
</div>
<div class="sc-aXZVg jGUWHs">
  <p>
    100
  </p>
  <p>
    #D9D9D9
  </p>
</div>
<div class="sc-aXZVg eWvEtr">
  <p>
    200
  </p>
  <p>
    #C5C5C5
  </p>
</div>
<div class="sc-aXZVg dYISWk">
  <p>
    300
  </p>
  <p>
    #B3B3B3
  </p>
</div>
<div class="sc-aXZVg hivEzf">
  <p>
    400
  </p>
  <p>
    #9F9F9F
  </p>
</div>
<div class="sc-aXZVg KQiML">
  <p>
    500
  </p>
  <p>
    #8C8C8C
  </p>
</div>
<div class="sc-aXZVg eqlNKi">
  <p>
    600
  </p>
  <p>
    #797979
  </p>
</div>
<div class="sc-aXZVg kWNRtA">
  <p>
    700
  </p>
  <p>
    #666666
  </p>
</div>
<div class="sc-aXZVg hKzxVe">
  <p>
    800
  </p>
  <p>
    #535353
  </p>
</div>
<div class="sc-aXZVg iQqaic">
  <p>
    900
  </p>
  <p>
    #404040
  </p>
</div>
`;

exports[`Building-Blocks/Colors Red smoke-test 1`] = `
<div class="sc-aXZVg boCnkt">
  <p>
    50
  </p>
  <p>
    #F7E7E6
  </p>
</div>
<div class="sc-aXZVg hNWjTh">
  <p>
    100
  </p>
  <p>
    #EFCFCE
  </p>
</div>
<div class="sc-aXZVg kJHZcZ">
  <p>
    200
  </p>
  <p>
    #E7B6B4
  </p>
</div>
<div class="sc-aXZVg cTWKCV">
  <p>
    300
  </p>
  <p>
    #DF9E9C
  </p>
</div>
<div class="sc-aXZVg cjRdeq">
  <p>
    400
  </p>
  <p>
    #D78683
  </p>
</div>
<div class="sc-aXZVg kthGIC">
  <p>
    500
  </p>
  <p>
    #CF6E6B
  </p>
</div>
<div class="sc-aXZVg fWohTI">
  <p>
    600
  </p>
  <p>
    #C75652
  </p>
</div>
<div class="sc-aXZVg jsOqwU">
  <p>
    700
  </p>
  <p>
    #BF3E3A
  </p>
</div>
<div class="sc-aXZVg bDGfJP">
  <p>
    800
  </p>
  <p>
    #B72620
  </p>
</div>
<div class="sc-aXZVg bCTxdW">
  <p>
    900
  </p>
  <p>
    #AF0E08
  </p>
</div>
`;
